File: 460-lfu-cache/lfu-cache.java
class LFUCache {
    Map<Integer, Pair<Integer, Integer>> cache; //key, (value, frequency) pair
    Map<Integer, LinkedHashSet<Integer>> freqs; //frequency, set of keys with same frequency
    int capacity;
    int minFreq;
    public LFUCache(int capacity) {
        cache = new HashMap<>();
        freqs = new HashMap<>();
        this.capacity = capacity;
        minFreq = 0;
    }
    
    public int get(int key) {
        //if key doesn't exist, return -1
        if (!cache.containsKey(key)) return -1;

        //get value, frequency
        int value = cache.get(key).getKey();
        int freq = cache.get(key).getValue();

        //update freq + 1

        //remove key from currrent freqs
        Set<Integer> keys = freqs.get(freq);
        keys.remove(key);
        if (keys.isEmpty()) {
            freqs.remove(freq);
            if (freq == minFreq) {
                minFreq++;
            }
        }

        //add back with +1 freq
        insert(key, value, freq + 1);
        return value;
    }
    
    public void put(int key, int value) {
        if (cache.containsKey(key)) {
            int freq = cache.get(key).getValue();
            get(key);
            cache.put(key, new Pair<>(value, freq + 1));
            return;
        }

        //new entry, remove least frequently used
        if (capacity == cache.size()) {
            Set<Integer> minKeys = freqs.get(minFreq);
            int toRemove = minKeys.iterator().next();
            minKeys.remove(toRemove);
            cache.remove(toRemove);
            if (minKeys.isEmpty()) {
                freqs.remove(minFreq);
            }
        }

        minFreq = 1;
        insert(key, value, 1);
    }

    private void insert(int key, int value, int freq) {
        freqs.putIfAbsent(freq, new LinkedHashSet<>());
        freqs.get(freq).add(key);
        cache.put(key, new Pair<>(value, freq));
    }
}

/**
 * Your LFUCache object will be instantiated and called as such:
 * LFUCache obj = new LFUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

File: 460-lfu-cache/lfu-cache.py
class LFUCache:

    def __init__(self, capacity: int):
        self.cache = {} # key: freq, value
        self.freqs = {} # freq: list of (keys: None)
        self.capacity = capacity
        self.min_freq = 0

    def get(self, key: int) -> int:
        if key not in self.cache:
            return -1
        freq, value = self.cache[key]
        del self.freqs[freq][key]
        if not self.freqs[freq]:
            del self.freqs[freq]
            if freq == self.min_freq:
                self.min_freq = freq + 1
        self.insert(key, value, freq + 1)
        return value
    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            self.cache[key] = (self.cache[key][0], value)
            self.get(key)
            return
        if self.capacity == len(self.cache):
            keysToRemove = self.freqs[self.min_freq]
            keyToRemove = next(iter(keysToRemove.keys()))
            del self.cache[keyToRemove]
            del self.freqs[self.min_freq][keyToRemove]
            if not self.freqs[self.min_freq]:
                del self.freqs[self.min_freq]
        self.min_freq = 1
        self.insert(key, value, 1)
    def insert(self, key, value, freq):
        self.cache[key] = (freq, value)
        if freq not in self.freqs:
            self.freqs[freq] = defaultdict()
        self.freqs[freq][key] = None


# Your LFUCache object will be instantiated and called as such:
# obj = LFUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)